Tests/GRPCTests/ClientConnectionBackoffTests.swift

```diff
@@ -18,82 +18,169 @@ import GRPC
+    self.client.connectivity.onNext(state: .ready) {
+      connectionReady.fulfill()
+    }
+
+    self.wait(for: [transientFailure], timeout: 1.0)
 
     self.server = try self.makeServer()
+    let serverStarted = self.expectation(description: "server started")
     self.server.assertSuccess(fulfill: serverStarted)
 
-    self.wait(for: [serverStarted, clientConnected], timeout: 2.0, enforceOrder: true)
+    self.wait(for: [serverStarted, connectionReady], timeout: 2.0, enforceOrder: true)
+    XCTAssertEqual(self.stateDelegate.states, [.connecting, .transientFailure, .connecting, .ready])
   }
 
   func testClientEventuallyTimesOut() throws {
-    var configuration = self.makeClientConfiguration()
-    configuration.connectionBackoff = ConnectionBackoff(maximumBackoff: 0.1)
+    let connectionShutdown = self.expectation(description: "connection shutdown")
+    self.client = self.makeClientConnection(self.makeClientConfiguration())
+    self.client.connectivity.onNext(state: .shutdown) {
+      connectionShutdown.fulfill()
+    }
 
-    self.client = self.makeClientConnection(configuration)
-    XCTAssertThrowsError(try self.client.wait()) { error in
-      XCTAssert(error is NIOConnectionError)
+    self.wait(for: [connectionShutdown], timeout: 1.0)
+    XCTAssertEqual(self.stateDelegate.states, [.connecting, .transientFailure, .connecting, .shutdown])
+  }
+
+  func testClientReconnectsAutomatically() throws {
+    self.server = try self.makeServer()
+    let server = try self.server.wait()
+
+    let connectionReady = self.expectation(description: "connection ready")
+    self.client = self.makeClientConnection(self.makeClientConfiguration())
+    self.client.connectivity.onNext(state: .ready) {
+      connectionReady.fulfill()
     }
+
+    // Once the connection ready we can kill the server.
+    self.wait(for: [connectionReady], timeout: 1.0)
+    XCTAssertEqual(self.stateDelegate.clearStates(), [.connecting, .ready])
+
+    try server.close().wait()
+    try self.serverGroup.syncShutdownGracefully()
+    self.server = nil
+    self.serverGroup = nil
+
+    let transientFailure = self.expectation(description: "connection transientFailure")
+    self.client.connectivity.onNext(state: .transientFailure) {
+      transientFailure.fulfill()
+    }
+
+    self.wait(for: [transientFailure], timeout: 1.0)
+    XCTAssertEqual(self.stateDelegate.clearStates(), [.connecting, .transientFailure])
+
+    let reconnectionReady = self.expectation(description: "(re)connection ready")
+    self.client.connectivity.onNext(state: .ready) {
+      reconnectionReady.fulfill()
+    }
+
+    // Start a new server.
+    self.serverGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    self.server = try self.makeServer()
+
+    self.wait(for: [reconnectionReady], timeout: 1.0)
+    XCTAssertEqual(self.stateDelegate.clearStates(), [.connecting, .ready])
+
+    // Ensure we can actually make a call.
```

Comment:
Could we also test making a call after the first server has shut down, but before the second one has been started?

Reply:
We can, but the reconnect will already be taking place: if we `wait()` on that call to complete (before starting the server) it will eventually time out the reconnect (and cause a `.shutdown`). 

If we don't `wait()` on it, then it will eventually connect and that call will succeed.

I'll add the former as a separate test.

Reply:
> If we don't wait() on it, then it will eventually connect and that call will succeed.

I think that's what we should do — start the call while the server is offline, then `wait()` for it to succeed _after_ the server is back up.